File: backend/ria_ml/CMULockboxes_process.m
function CMULockboxes = CMULockboxes_process(CMULockboxes, market, client );
  % computes lockbox proportions for an CMULockbox strategy
  
  % get number of years
     [nscen nyrs] = size(market.cumRmsM);
     
  % set proportions for year 1
     mktprop = CMULockboxes.initialMarketProportion;
     if mktprop > 1; mktprop = 1; end;
     if mktprop < 0; mktprop = 0; end;
     tipsprop =  1 - mktprop;
     
  % find ratio of market proportion each year to that for the prior year
     a = market.avec(2);
     b = market.b;
     logk = ( -log(1/a) ) / b;
     k = exp(logk);
  % compute market proportions for all years  
     mktprops = mktprop* ( (1/k).^(0:1:nyrs-1) );
     
  % compute TIPS proportions for all years   
     tipsprops = tipsprop * ( (1/market.rf).^(0:1:nyrs-1) );
     
  % compute lockbox proportions;
      CMULockboxes.proportions = [ tipsprops; mktprops ];

  % plot contents if requested
     if lower(CMULockboxes.showProportions) == 'y'
        xs =  CMULockboxes.proportions; 
        fig = figure;
        x = 1:1:size(xs,2);
        bar(x,xs','stacked'); grid;
        set(gca,'FontSize',30);
        ss = client.figurePosition;
        set(gcf,'Position',ss);
        set(gcf,'Color',[1 1 1]);
        xlabel('Lockbox Maturity Year ','fontsize',30);
        ylabel('Amount Invested at Inception   ','fontsize',30);
        legend('TIPS ','Market ');
        ax = axis; ax(1) = 0; ax(2) = nyrs+1; ax(3) = 0; ax(4) = 1; axis(ax);
        t = ['Lockbox Proportions for Constant Marginal Utility '];
        title(t,'Fontsize',40,'Color','b');
        beep; pause;
     end; %if lower(CMULockboxes.showContents) = 'y'    
     
end % function 

File: backend/ria_ml/CMULockboxes_process.py
import numpy as np
import matplotlib.pyplot as plt

def CMULockboxes_process(CMULockboxes, market, client):
    nscen, nyrs = market['cumRmsM'].shape

    mktprop = np.clip(CMULockboxes['initialMarketProportion'], 0, 1)
    tipsprop = 1 - mktprop

    a = market['avec'][1]
    b = market['b']
    logk = -np.log(1 / a) / b
    k = np.exp(logk)
    mktprops = mktprop * ((1 / k)**np.arange(nyrs))
    tipsprops = tipsprop * ((1 / market['rf'])**np.arange(nyrs))

    CMULockboxes['proportions'] = np.vstack([tipsprops, mktprops])

    if CMULockboxes['showProportions'].lower() == 'y':
        xs = CMULockboxes['proportions']
        fig, ax = plt.subplots()
        x = np.arange(1, xs.shape[1] + 1)
        ax.bar(x, xs[0, :], label='TIPS')
        ax.bar(x, xs[1, :], bottom=xs[0, :], label='Market')
        ax.grid(True)
        ax.set_xlabel('Lockbox Maturity Year')
        ax.set_ylabel('Amount Invested at Inception')
        ax.legend(['TIPS', 'Market'])
        ax.set_xlim([0, nyrs + 1])
        ax.set_title('Lockbox Proportions for Constant Marginal Utility', color='b')
        plt.show()

    return CMULockboxes
